File: 03_frontend/types/competition.ts
/**
 * コンペティション関連の型定義
 */

/**
 * 評価指標の型
 */
export interface EvaluationMetric {
  metric: string;
  explanation: string;
  why_important: string;
}

/**
 * 構造化要約の型
 */
export interface StructuredSummary {
  overview: string;
  objective: string;
  data: string;
  evaluation: EvaluationMetric;
  business_value: string;
  key_challenges: string[];
}

/**
 * データセットのカラム情報
 */
export interface DatasetColumn {
  name: string;
  description: string;
}

/**
 * データセット情報の型
 */
export interface DatasetInfo {
  files: string[];
  total_size: string;
  description: string;
  features: string[];
  columns: DatasetColumn[];
}

export interface Competition {
  id: string;
  title: string;
  url: string;
  start_date: string | null;
  end_date: string | null;
  status: 'active' | 'completed';
  metric: string | null;
  metric_description?: string | null;
  description?: string;
  summary: string | null; // 日本語要約テキスト
  tags: string[]; // JSON配列
  data_types: string[]; // JSON配列
  competition_features: string[]; // JSON配列
  task_types: string[]; // JSON配列
  domain: string | null;
  dataset_info?: string | null;
  discussion_count: number;
  solution_status: '未着手' | 'ディスカッションのみ' | '解法分析済み';
  is_favorite: boolean;
  days_until_deadline?: number | null; // 終了日までの日数（開催中のみ）
  created_at: string;
}

export interface CompetitionListResponse {
  items: Competition[];
  total: number;
  active_count: number;
  completed_count: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface CompetitionQueryParams {
  page?: number;
  limit?: number;
  status?: string;
  search?: string;
  sort_by?: string;
  order?: 'asc' | 'desc';
}

/**
 * ディスカッション情報の型
 */
export interface Discussion {
  id: number;
  competition_id: string;
  title: string;
  author: string | null;
  author_tier: string | null;
  tier_color: string | null;
  url: string;
  vote_count: number;
  comment_count: number;
  view_count: number;
  kaggle_created_at: string | null;
  category: string | null;
  is_pinned: boolean;
  content: string | null;
  summary: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * 解法情報の型
 */
export interface Solution {
  id: number;
  competition_id: string;
  title: string;
  author: string | null;
  author_tier: string | null;
  tier_color: string | null;
  url: string;
  type: 'notebook' | 'discussion';
  medal: 'gold' | 'silver' | 'bronze' | null;
  rank: number | null;
  vote_count: number;
  comment_count: number;
  content: string | null;
  summary: string | null;
  techniques: string | null;
  created_at: string;
  updated_at: string;
}

File: 02_backend/app/models/competition.py
"""
Competitionモデル
"""
from dataclasses import dataclass, field, asdict
from datetime import datetime
from typing import Optional, List, Dict, Any


@dataclass
class Competition:
    """コンペティション情報"""

    # 必須フィールド
    id: str
    title: str
    url: str
    status: str  # 'active', 'completed', 'upcoming'

    # オプショナルフィールド
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    metric: Optional[str] = None
    metric_description: Optional[str] = None
    description: Optional[str] = None
    summary: Optional[str] = None
    tags: List[str] = field(default_factory=list)
    data_types: List[str] = field(default_factory=list)
    competition_features: List[str] = field(default_factory=list)
    task_types: List[str] = field(default_factory=list)
    domain: Optional[str] = None
    dataset_info: Optional[str] = None  # JSON文字列
    discussion_count: int = 0
    solution_status: str = "未着手"
    is_favorite: bool = False

    # メタデータ
    created_at: Optional[datetime] = None
    last_scraped_at: Optional[datetime] = None

    def to_dict(self) -> Dict[str, Any]:
        """Dictに変換"""
        data = asdict(self)

        # datetimeをISO形式の文字列に変換
        for key in ['start_date', 'end_date', 'created_at', 'last_scraped_at']:
            if data.get(key) and isinstance(data[key], datetime):
                data[key] = data[key].isoformat()

        return data

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Competition':
        """Dictから作成"""
        # datetimeフィールドを変換
        datetime_fields = ['start_date', 'end_date', 'created_at', 'last_scraped_at']
        for field_name in datetime_fields:
            if data.get(field_name) and isinstance(data[field_name], str):
                try:
                    data[field_name] = datetime.fromisoformat(data[field_name])
                except (ValueError, TypeError):
                    data[field_name] = None

        # リストフィールドのデフォルト値
        if 'tags' not in data:
            data['tags'] = []
        if 'data_types' not in data:
            data['data_types'] = []
        if 'competition_features' not in data:
            data['competition_features'] = []
        if 'task_types' not in data:
            data['task_types'] = []

        return cls(**{k: v for k, v in data.items() if k in cls.__dataclass_fields__})
